_posts/2022-05-23-deploy-agent-on-fedora-coreos.markdown

```diff
@@ -0,0 +1,314 @@
+- `protocol`: set to `http`. 
+- `server`: address of the `Flotta Edge API`. The domain name has to be the same as the domain set in the cluster because the certificates are valid only for that domain.
+
+The certificates can be obtain by running `make get-certs` in `Flotta Operator` cluster.
+
+There is one line missing here, which is `client-id`. We will add later on this line cause we need the `machine-id` and this is available only after the first boot.
+
+## Creating of the ignition file
+
+CoreOS configure the system using a [ignition](https://docs.fedoraproject.org/en-US/fedora-coreos/producing-ign/) file. The `ignition` file is produced from a `butane` file.
+You can find the specification of the `butane` file on [coreos/butane](https://coreos.github.io/butane/config-fcos-v1_3/). Basically, the `butane` file is a `yaml` file which specify
+how the system will be configured after startup. 
+
+All the system configuration is done via `systemd` so to install and setup `flotta-agent` we need to create two services, one for install and one for setup. These services needs to run only one time when we are provisioning the system.
+
+### Install `flotta-agent`
+
+Installing the `flotta-agent` through `systemd` is easy. We need to run install the `copr` and to run `rpm-ostree`, but we need to install only one time and we need to reboot the OS after the install. 
+
+**Remark**: CoreOS is a `ostree` OS hence it is using `rpm-ostree` instead of `dnf`.
+
+The service could look like this:
+```yaml
+[Unit]
+Description=Flotta setup service
+Wants=network-online.target
+After=network-online.target
+ConditionPathExists=!/usr/lib/systemd/system/yggdrasild.service
+
+[Service]
+Type=oneshot
+ExecStartPre=curl https://copr.fedorainfracloud.org/coprs/project-flotta/flotta/repo/fedora-35/project-flotta-flotta-fedora-35.repo -o /etc/yum.repos.d/project-flotta.repo
+ExecStart=rpm-ostree install node_exporter yggdrasil flotta-agent
+ExecStart=/bin/bash -c "echo \"client-id = ##`cat /etc/machine-id`##\" >> /etc/yggdrasil/config.toml"
+ExecStart=sed -i -s "s/##/\"/g" /etc/yggdrasil/config.toml
+ExecStopPost=systemctl reboot
+
+[Install]
+WantedBy=multi-user.target
+```
+
+For those not familiar with `systemd`, here are some explications:
+
+- the service must start after the `network-online.target` meaning we need to have an internet connection.
+- `ConditionPathExists` must be true for this service to start. This condition allows us to start this service only if `yggdrasild.service` has not been install yet.
+- Before actually install the `flotta-agent`, we need to setup the `copr`. This is done with the target `ExecStartPre`
+- `ExecStart` is installing what we need. The next `ExecStart` targets are setting the `client-id`. Basically, what we what is to modify the provided `yggdrasil` configuration file to set the `client-id` to the `machine-id`. 
```

Comment:
Not sure: s/is installing/installs/

Since there are multiple occurrences of ``ExecStart`, WDYT to describe them explicitly (i.e. line by line)? Will it be too much in your opinion?

Reply:
installs I think